packages/bull-vault/src/AuctionBull.sol

```diff
@@ -0,0 +1,565 @@
+            }
+
+            _rebalanceLeverageComponent(_wethTargetInEuler, _wethLimitPrice);
+        }
+    }
+
+    function _executeCrabDeposit(uint256 _crabAmount, uint256 _wethTargetInEuler, uint256 _wethLimitPrice, uint256 _ethInCrab) internal {
+        uint256 wethFromEuler;
+        uint256 ethNeededForCrab = _crabAmount.wdiv(IERC20(crab).totalSupply()).wmul(_ethInCrab);
+        uint256 wethInCollateral = IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy));
+        if (_wethTargetInEuler > wethInCollateral) {
+            uint256 wethToBuy;
+            uint256 usdcToBorrow;
+            {
+                uint256 ethUsdPrice = UniOracle._getTwap(ethUSDCPool, weth, usdc, TWAP, false);
+                wethToBuy = _wethTargetInEuler.sub(wethInCollateral);
+                usdcToBorrow = wethToBuy.wmul(ethUsdPrice).div(1e12);
+            }
+            _exactOutFlashSwap(
+                usdc,
+                weth,
+                3000,
+                wethToBuy,
+                usdcToBorrow.wmul(_wethLimitPrice),
+                uint8(FLASH_SOURCE.FULL_REBALANCE_BORROW_USDC_BUY_WETH),
+                abi.encodePacked(wethToBuy, usdcToBorrow)
+            );
+        } else {
+            wethFromEuler = wethInCollateral.sub(_wethTargetInEuler);
+            IBullStrategy(bullStrategy).repayAndWithdrawFromLeverage(0, wethFromEuler);
+        }
+        ethNeededForCrab = ethNeededForCrab.sub(wethFromEuler);
+        IBullStrategy(bullStrategy).depositEthIntoCrab(ethNeededForCrab);
+    }
+
+    function _rebalanceLeverageComponent(uint256 _wethTargetInEuler, uint256 _wethLimitPrice) internal {
+        uint256 remainingWeth = IERC20(weth).balanceOf(address(this));
+        uint256 wethInCollateral = IEulerEToken(eToken).balanceOfUnderlying(address(bullStrategy));
+        if (_wethTargetInEuler > remainingWeth.add(wethInCollateral)) {
+            uint256 ethUsdPrice = UniOracle._getTwap(ethUSDCPool, weth, usdc, TWAP, false);
+            uint256 wethToBuy = _wethTargetInEuler.sub(remainingWeth.add(wethInCollateral));
+            uint256 usdcToBorrow = wethToBuy.wmul(ethUsdPrice).div(1e12);
+            _exactOutFlashSwap(
+                usdc,
+                weth,
+                3000,
+                wethToBuy,
+                usdcToBorrow.wmul(_wethLimitPrice),
+                uint8(FLASH_SOURCE.FULL_REBALANCE_BORROW_USDC_BUY_WETH),
+                abi.encodePacked(wethToBuy, usdcToBorrow)
+            );
+        } else {
+            uint256 wethToSell = remainingWeth.add(wethInCollateral).sub(_wethTargetInEuler);
+            uint256 usdcToRepay = _exactInFlashSwap(
+                weth,
+                usdc,
+                3000,
+                wethToSell,
+                wethToSell.wdiv(_wethLimitPrice).div(1e12),
+                uint8(FLASH_SOURCE.GENERAL_SWAP),
+                ""
+            );
+
+            IBullStrategy(bullStrategy).repayAndWithdrawFromLeverage(usdcToRepay, 0);
```

Comment:
its likely that we need to withdraw eth from euler to pay the uniswap flashswap back, so think this needs to be in the flashswap callback so it happens before we have to pay the weth to uniswap back

usdcToRepay should be the amount we repay, which you do, but we will need to withdraw ```wethInCollateral- _wethTargetInEuler``` from euler.

Reply:
fixed